src/sql/plot.py

```diff
@@ -352,37 +474,107 @@ def histogram(payload, table, column, bins, with_=None, conn=None):
 
+    # FIXME: we're computing all the with elements twice
+    min_, max_ = _min_max(conn, table, column, with_=with_)
+
+    filter_query = f"WHERE {facet['key']} == '{facet['value']}'" if facet else ""
+
+    if _are_numeric_values(min_, max_):
+        if not isinstance(bins, int):
+            raise ValueError(
+                f"bins are '{bins}'. Please specify a valid number of bins."
+            )
+
+        range_ = max_ - min_
+        bin_size = range_ / bins
+
+        template = Template(
+            """
+            select
+            floor("{{column}}"/{{bin_size}})*{{bin_size}},
+            count(*) as count
+            from "{{table}}"
+            {{filter_query}}
+            group by 1
+            order by 1;
+            """
+        )
+        query = template.render(
+            table=table, column=column, bin_size=bin_size, filter_query=filter_query
+        )
+    else:
+        template = Template(
+            """
+        select
+            "{{column}}", count ({{column}})
+        from "{{table}}"
+        {{filter_query}}
+        group by 1
+        order by 1;
+        """
+        )
+        query = template.render(table=table, column=column, filter_query=filter_query)
+
+    if with_:
+        query = str(store.render(query, with_=with_))
+
+    query = sql.connection.Connection._transpile_query(query)
+    data = conn.execute(query).fetchall()
+    bin_, height = zip(*data)
+
+    if bin_[0] is None:
+        raise ValueError("Data contains NULLs")
+
+    return bin_, height
+
+
+@modify_exceptions
+def _histogram_stacked(
+    table, column, category, bins, with_=None, conn=None, facet=None
+):
+    """Compute bin size for the column and the corresponding
+    heights of each category value"""
+    if not conn:
+        conn = sql.connection.Connection.current.session
+
     # FIXME: we're computing all the with elements twice
     min_, max_ = _min_max(conn, table, column, with_=with_)
     range_ = max_ - min_
     bin_size = range_ / bins
 
+    filter_query = f"WHERE {facet['key']} == '{facet['value']}'" if facet else ""
+
     template = Template(
         """
 select
   floor("{{column}}"/{{bin_size}})*{{bin_size}},
-  count(*) as count
+  {{category}}
 from "{{table}}"
-group by 1
-order by 1;
+{{filter_query}}
 """
     )
-
-    query = template.render(table=table, column=column, bin_size=bin_size)
+    query = template.render(
+        table=table,
+        column=column,
+        bin_size=bin_size,
+        category=category,
+        filter_query=filter_query,
+    )
 
     if with_:
         query = str(store.render(query, with_=with_))
 
-    query = sql.connection.Connection._transpile_query(query)
-    data = conn.execute(query).fetchall()
-    bin_, height = zip(*data)
+    # TODO: We can use the duckdb pivot feature which will be released
+    # in 0.8 so we won't have to use dataframe
+    # https://github.com/duckdb/duckdb/pull/6387
+    import pandas as pd
```

Comment:
since matplotlib already uses numpy as dependency, maybe we can use that? https://numpy.org/doc/stable/reference/generated/numpy.transpose.html

if we need to keep it, then let's add pandas to the requires decorator (as mentioned in my previous comment). also, let's move this import to the top and wrap it in a try except statement.

Reply:
and let's also add an admonition in the tutorial saying that ggplot requires matplotlib (and pandas if we decide to keep it)